staging/src/k8s.io/apimachinery/pkg/util/cache/lruexpirecache.go

```diff
@@ -61,29 +61,88 @@ func NewLRUExpireCacheWithClock(maxSize int, clock Clock) *LRUExpireCache {
 type cacheEntry struct {
 	value      interface{}
 	expireTime time.Time
+
+	// If nil, a computation is in progress. You must wait and then re-get
+	// the cache entry.
+	wait <-chan struct{}
 }
 
 // Add adds the value to the cache at key with the specified maximum duration.
 func (c *LRUExpireCache) Add(key interface{}, value interface{}, ttl time.Duration) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	c.cache.Add(key, &cacheEntry{value, c.clock.Now().Add(ttl)})
+	c.cache.Add(key, &cacheEntry{value, c.clock.Now().Add(ttl), nil})
 }
 
 // Get returns the value at the specified key from the cache if it exists and is not
 // expired, or returns false.
 func (c *LRUExpireCache) Get(key interface{}) (interface{}, bool) {
+	return c.getCounted(key, 0)
+}
+
+func (c *LRUExpireCache) getCounted(key interface{}, count int) (interface{}, bool) {
+	if count >= 2 {
+		return nil, false
+	}
+	ce, _ := c.getOrBegin(key, 0)
+	if ce == nil {
+		return nil, false
+	}
+	return c.finishGet(key, ce, count)
+}
+
+func (c *LRUExpireCache) finishGet(key interface{}, ce *cacheEntry, count int) (interface{}, bool) {
+	if ce.wait == nil {
+		return ce.value, true
+	}
+	select {
+	case <-ce.wait:
+		return c.getCounted(key, count+1)
+	case <-time.After(time.Until(ce.expireTime)):
+		return nil, false
+	}
+}
+
+// GetOrWait returns the value at the specified key from the cache if it exists
+// and is not expired. Otherwise, it calls the compute() function to produce
+// the value; requests for the same key that arrive within computationTime will
+// be blocked while the computation is in progress; for all such requests,
+// compute() will only be called once.
+func (c *LRUExpireCache) GetOrWait(key interface{}, compute func() (value interface{}, ttl time.Duration), computationTime time.Duration) (interface{}, bool) {
+	ce, ch := c.getOrBegin(key, computationTime)
+	if ch != nil {
+		go func() {
+			defer close(ch)
+			value, ttl := compute()
```

Comment:
Maybe utilize a pool of workers instead of forking goroutine every time.

Reply:
Why? goroutines do not need to be conserved and this is vastly simpler.